### PR TITLE
String type for DiemID

### PR DESCRIPTION
**Pull request by sunmilee**
## Motivation

Creating a new string type that enforces a character length limit as well as a character set limit as defined by DIP-10.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?
Yes

## Test Plan

diem_id_identifier_tests.rs
cargo x test -p diem-types

### CODE REVIEW

File: types/src/transaction/diem_id_identifier.rs [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+            Some('a'..='z') | Some('A'..='Z') | Some('0'..='9') => {}
+            Some(_) => return Err(DiemIdParseError),
+            None => return Err(DiemIdParseError),
+        }
+        for c in chars {
+            match c {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '.' => {}
+                _ => return Err(DiemIdParseError),
+            }
+        }
```
Comment (davidiw) [inline types/src/transaction/diem_id_identifier.rs:95]: add some new lines for readability

File: types/src/transaction/diem_id_identifier.rs [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+        let mut chars = s.chars();
+        match chars.next() {
+            Some('a'..='z') | Some('A'..='Z') | Some('0'..='9') => {}
+            Some(_) => return Err(DiemIdParseError),
+            None => return Err(DiemIdParseError),
+        }
+        for c in chars {
+            match c {
+                'a'..='z' | 'A'..='Z' | '0'..='9' | '.' => {}
+                _ => return Err(DiemIdParseError),
```
Comment (davidiw) [inline types/src/transaction/diem_id_identifier.rs:93]: should we return back the index of the invalid character?

File: types/src/transaction/diem_id_identifier.rs [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
```
Comment (davidiw) [inline types/src/transaction/diem_id_identifier.rs:133]: return back index of invalid character?

File: types/src/transaction/diem_id_identifier.rs [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+        DiemIdUserIdentifier::new(&s).map_err(D::Error::custom)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Serialize)]
+pub struct DiemIdVaspDomainIdentifier(Box<str>);
+
+impl DiemIdVaspDomainIdentifier {
+    pub fn new(s: &str) -> Result<Self, DiemIdParseError> {
+        if s.len() > 63 || s.is_empty() {
```
Comment (davidiw) [inline types/src/transaction/diem_id_identifier.rs:121]: whitespace below?

**Review by davidiw — COMMENTED** [review]
overall seems good, but any reason to shove this under transactions versus types?

File: types/src/diem_id_identifier.rs [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DiemIdParseError {
+    message: String,
+}
+
+impl DiemIdParseError {
+    pub fn new(message: String) -> DiemIdParseError {
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:184]: Let's not mark this as public (as a way to hide the implementation details of the Error type)

File: types/src/diem_id_identifier.rs [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+    type Err = DiemIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, DiemIdParseError> {
+        if !s.contains('@') {
+            return Err(DiemIdParseError::new("DiemId does not have @".to_string()));
+        };
+        let split = s.split('@');
+        let vec = split.collect::<Vec<&str>>();
+        DiemId::new_from_raw(&vec[0], &vec[1])
+    }
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:73]: This is close but I think there's a few corner cases that still need to be covered (and adding more negative tests for this would be nice). For example if you have the string `bmwill@novi@foo` this implementation would accept that string, silently dropping the `@foo` part when in reality we'd want to error out on that. You should be able to fix that by using a method like [split_once](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_once) but since that's not available in rust 1.51 you should be able to use a combination of [find](https://doc.rust-lang.org/stable/std/primitive.str.html#method.find) and [split_at](https://doc.rust-lang.org/stable/std/primitive.str.html#method.split_at).

File: types/src/diem_id_identifier.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+        user_identifier: DiemIdUserIdentifier,
+        vasp_domain_identifier: DiemIdVaspDomainIdentifier,
+    ) -> Self {
+        Self {
+            user_identifier,
+            vasp_domain_identifier,
+        }
+    }
+
+    pub fn new_from_raw(
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:33]: Lets mark this function as private for now.

File: types/src/diem_id_identifier.rs [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+    {
+        use serde::de::Error;
+
+        let s = <String>::deserialize(deserializer)?;
+        DiemIdVaspDomainIdentifier::new(&s).map_err(D::Error::custom)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct DiemIdParseError {
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:179]: Can you add an `impl std::error::Error for DiemIdParseError {}`?

File: types/src/lib.rs [hunk 9]
```diff
@@ -3,6 +3,8 @@
 
 #![forbid(unsafe_code)]
 
+pub use account_address::AccountAddress as PeerId;
```
Comment (bmwill) [inline types/src/lib.rs:6]: This change seems un related to this PR?

File: types/src/transaction/mod.rs [hunk 10]
```diff
@@ -1,6 +1,36 @@
 // Copyright (c) The Diem Core Contributors
 // SPDX-License-Identifier: Apache-2.0
 
+use std::{
```
Comment (bmwill) [inline types/src/transaction/mod.rs:4]: This change seems unrelated to this PR?

**Review by bmwill — COMMENTED** [review]
Almost there! Just a few last things and maybe some more negative testing and we should be good to go.

File: types/src/diem_id_identifier.rs [hunk 11]
```diff
@@ -0,0 +1,289 @@
...
+    fn from_str(s: &str) -> Result<Self, DiemIdParseError> {
+        let index = s
+            .find('@')
+            .ok_or_else(|| DiemIdParseError::new("DiemId does not have @".to_string()))?;
+        let split = s.split_at(index);
+        DiemId::new_from_raw(split.0, (split.1).trim_start_matches('@'))
+    }
+}
+
+impl std::error::Error for DiemIdParseError {}
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:75]: Can you move this closer to the definition of `DiemIdParseError`?

File: types/src/diem_id_identifier.rs [hunk 12]
```diff
@@ -0,0 +1,289 @@
...
+
+impl FromStr for DiemId {
+    type Err = DiemIdParseError;
+
+    fn from_str(s: &str) -> Result<Self, DiemIdParseError> {
+        let index = s
+            .find('@')
+            .ok_or_else(|| DiemIdParseError::new("DiemId does not have @".to_string()))?;
+        let split = s.split_at(index);
+        DiemId::new_from_raw(split.0, (split.1).trim_start_matches('@'))
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:71]: The `trim_start_matches` function results in `bmwill@@novi` being able to pass the validation. Lets instead change this to 
```suggestion
        DiemId::new_from_raw(split.0, &(split.1)[1..])
```

Comment (bmwill) [reply types/src/diem_id_identifier.rs:71]: Lets also add a test case for this example

File: types/src/diem_id_identifier.rs [hunk 13]
```diff
@@ -0,0 +1,289 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+use serde::{de, Serialize};
+use std::{
+    fmt::{Display, Formatter},
+    str::FromStr,
+};
+
+/// This file implements length and character set limited string types needed
+/// for DiemID as defined by DIP-10: https://github.com/diem/dip/blob/main/dips/dip-10.md
```
Comment (bmwill) [inline types/src/diem_id_identifier.rs:11]:
```suggestion
//! This file implements length and character set limited string types needed
//! for DiemID as defined by DIP-10: https://github.com/diem/dip/blob/main/dips/dip-10.md
```

And maybe put this above the `use` statements.

**Review by bmwill — COMMENTED** [review]
So close! Really just 1 small tweak left and we can get this landed.

File: types/src/diem_id_identifier.rs (hunk 11 above)
Comment (sunmilee) [reply types/src/diem_id_identifier.rs:75]: lol not sure why i wrote it there

**Review by bmwill — APPROVED** [review]
Looks great!
